File: frontend/tool/extract_login_arb.py
"""One-off: parse login_screen.dart _copy maps -> ARB key list. Run from frontend/: python tool/extract_login_arb.py"""
import re
from pathlib import Path

text = Path("lib/screens/login_screen.dart").read_text(encoding="utf-8")


def parse_block(locale: str) -> dict[str, str]:
    start = text.index(f"'{locale}': {{") + len(f"'{locale}': {{")
    depth = 1
    i = start
    while i < len(text) and depth:
        if text[i] == "{":
            depth += 1
        elif text[i] == "}":
            depth -= 1
        i += 1
    block = text[start : i - 1]
    out: dict[str, str] = {}
    # multiline values: 'key': '...', possibly spanning lines
    for m in re.finditer(
        r"'([^']+)':\s*((?:'(?:\\'|[^'])*'|\"(?:\\\"|[^\"])*\"))",
        block,
        re.S,
    ):
        k, raw = m.group(1), m.group(2)
        if raw.startswith("'"):
            v = raw[1:-1].replace("\\'", "'")
        else:
            v = raw[1:-1].replace('\\"', '"')
        out[k] = v
    return out


def to_arb_key(k: str) -> str:
    parts = k.split("_")
    return "login" + "".join(p[:1].upper() + p[1:] if p else "" for p in parts)


he = parse_block("he")
en = parse_block("en")
ru = parse_block("ru")
keys = sorted(he.keys())
assert set(keys) == set(en.keys()) == set(ru.keys()), (
    set(he.keys()) ^ set(en.keys()),
    set(he.keys()) ^ set(ru.keys()),
)

for loc, name in [(en, "en"), (he, "he"), (ru, "ru")]:
    lines = []
    for k in keys:
        ak = to_arb_key(k)
        val = loc[k].replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")
        lines.append(f'  "{ak}": "{val}"')
    Path(f"tool/_login_arb_{name}.txt").write_text(",\n".join(lines) + "\n", encoding="utf-8")
print("keys", len(keys))

File: frontend/tool/extract_login_arb.mjs
import fs from "fs";
const text = fs.readFileSync("lib/screens/login_screen.dart", "utf8");
function parseBlock(locale) {
  const startMark = `'${locale}': {`;
  const start = text.indexOf(startMark);
  if (start < 0) throw new Error("no " + locale);
  let i = start + startMark.length;
  let depth = 1;
  while (i < text.length && depth > 0) {
    if (text[i] === "{") depth++;
    else if (text[i] === "}") depth--;
    i++;
  }
  const block = text.slice(start + startMark.length, i - 1);
  const out = {};
  const re = /'([^']+)':\s*((?:'(?:\\'|[^'])*'|"(?:\\"|[^"])*"))/gs;
  let m;
  while ((m = re.exec(block))) {
    const k = m[1];
    let raw = m[2];
    let v;
    if (raw.startsWith("'")) v = raw.slice(1, -1).replace(/\\'/g, "'");
    else v = raw.slice(1, -1).replace(/\\"/g, '"');
    out[k] = v;
  }
  return out;
}
function toArbKey(k) {
  return (
    "login" +
    k.split("_").map((p) => p.charAt(0).toUpperCase() + p.slice(1)).join("")
  );
}
function esc(s) {
  return s
    .replace(/\\/g, "\\\\")
    .replace(/"/g, '\\"')
    .replace(/\n/g, "\\n")
    .replace(/\r/g, "\\r");
}
const he = parseBlock("he");
const en = parseBlock("en");
const ru = parseBlock("ru");
const keys = Object.keys(he).sort();
for (const loc of ["en", "he", "ru"]) {
  const map = { en, he, ru }[loc];
  const lines = keys.map((k) => `  "${toArbKey(k)}": "${esc(map[k])}"`);
  fs.writeFileSync(`tool/_login_arb_${loc}.txt`, lines.join(",\n") + "\n");
}
console.log("keys", keys.length);
